### ISSUE TITLE
iFrame not returning .text

### ISSUE THREAD

**Issue by randyr90s** [watir/watir-webdriver#413]
I'm accessing a site where, for some reason, the iFrame won't fully load by itself in a browser tab...it needs the parent around it.

The .text method is returning an empty string on the iFrame:

ifrm = browser.iframe(:index => 0)
print "iframe html is " + ifrm.html      # works
print "iframe text is " + ifrm.text      # doesn't work, empty string

**Comment by randyr90s** [comment]
A 'hack' workaround is to save the .html to a local file, .goto that, then use .text.  

Ugly, but it's working.

**Comment by titusfortner** [comment]
`browser.iframe(:index => 0)` is an iframe element in the top level browsing context. Within the scope of the top level browsing context it does not have any text inside it, because it is opaque. To display a value, you need to switch into the iframe's browsing context. In Selenium you must do that explicitly, in Watir it does it implicitly so long as you are working with elements: `browser.iframe(index: 0).div(class: foo).text`

In your case, this will switch you into the iframe context and provide what you are looking for:
```
iframe_html = browser.iframe(index: 0).element.html
iframe_text = browser.iframe(index: 0).element.text
```
@p0deje - what do you think about making defining the browser methods in the FramedDriver class such that it automatically switches into the browsing context for these methods directly?

**Comment by randyr90s** [comment]
Thank you both for your quick replies.  I'm not quite sure I completely follow, however, as I thought
ifrm = browser.iframe(:index => 0)
would give me the iframe's content.

I'm also unclear why
browser.iframe(:index => 0).html
gives me results, yet
browser.iframe(:index => 0).text
does not.

**Comment by titusfortner** [comment]
IFrames are DOMs within the DOM. The driver can not see inside a nested DOM without being specifically told to switch into it. Right now Watir is currently implemented to only do that if you specify an element to interact with as a child element of the IFrame

So:
`browser.iframe(index: 0)` gives you a `Watir::IFrame` object, and there is no reason at that point for Watir to instruct the driver to switch into its context. 

Ah, I didn't realize `#html` is [already being treated special](https://github.com/watir/watir-webdriver/blob/master/lib/watir-webdriver/elements/iframe.rb#L26-L32).

I don't like the way it is being done, and we should have `#text` as an option as well.
If @p0deje likes #416 then we can give you that functionality in the next release.

**Comment by titusfortner** [comment]
fixed here - 6f55c8b2f463d5151b4cbacadd021685ed41ec70
